Organize remaining RemoteLayerTreeTransaction members

ruthvikkonda opened the pull request:
#### 5001daca24e7c29db01073cc27011c9928abc781
<pre>
Organize remaining RemoteLayerTreeTransaction members
<a href="https://bugs.webkit.org/show_bug.cgi?id=302267">https://bugs.webkit.org/show_bug.cgi?id=302267</a>
<a href="https://rdar.apple.com/164407141">rdar://164407141</a>

Reviewed by Matt Woodrow.

Right now RemoteLayerTreeDrawingAreaProxy::commitLayerTree currently receives a Vector&lt;std::pair&lt;RemoteLayerTreeTransaction,
RemoteScrollingCoordinatorTransaction&gt;&gt;, one pair per root frame. However, some aspects of the transaction are
scoped to the root frame. Some are scoped to the page. And some are scoped to the main frame. We want to accordingly
introduce some separation/organization here to reduce duplication and remove unnecessary isMainFrameProcessTransaction() checks.

This PR builds on <a href="https://github.com/WebKit/WebKit/pull/52671">https://github.com/WebKit/WebKit/pull/52671</a> and <a href="https://github.com/WebKit/WebKit/pull/52839.">https://github.com/WebKit/WebKit/pull/52839.</a> In this PR, we move
the last relevant members from transaction scope to main frame, page, and bundle scope accordingly. WCP propagation
is adjusted to fit this. And where appropriate, processing on the UIProcess side is separated to ensure that mainframe specific
data is only processed once rather than per transaction.

We should only be using bundle.mainFrameData now as the sole source of truth for checking for main frame identity, so isMainFrameProcessTransaction()
is removed. Furthermore, since these members got migrated, some logging was removed from RemoteLayerTreeTransaction. We add it back
in RemoteLayerTreeCommitBundle.mm.

No new functionality is added. This is a simple refactoring. Existing test coverage is enough.
As a result, no new tests are added.

* Source/WebCore/platform/ProcessQualified.h:
(WebCore::ProcessQualified::lessThanSameProcess const):
(WebCore::ProcessQualified::lessThanOrEqualSameProcess const):
(WebCore::ProcessQualified::greaterThanSameProcess const):
(WebCore::ProcessQualified::greaterThanOrEqualSameProcess const):
(WebCore::ProcessQualified::lessThanSameProcess): Deleted.
(WebCore::ProcessQualified::lessThanOrEqualSameProcess): Deleted.
(WebCore::ProcessQualified::greaterThanSameProcess): Deleted.
(WebCore::ProcessQualified::greaterThanOrEqualSameProcess): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.mm: Added.
(WebKit::PageData::description const):
(WebKit::MainFrameData::description const):
(WebKit::RemoteLayerTreeCommitBundle::description const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
(WebKit::RemoteLayerTreeTransaction::RemoteLayerTreeTransaction): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTreeDuringAnimatedResize:mainFrameData:]):
(-[WKWebView _trackTransactionCommit:]):
(-[WKWebView _updateScrollViewForTransaction:]):
(-[WKWebView _restoreScrollAndZoomStateForTransaction:]):
(-[WKWebView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(-[WKWebView _didCommitLayerTreeDuringAnimatedResize:]): Deleted.
(-[WKWebView _didCommitLayerTree:mainFrameData:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
(WebKit::PageClientImplCocoa::didCommitMainFrameData):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLayerTree):
(WebKit::PageClientImpl::didCommitMainFrameData):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(-[WKContentView _didCommitLayerTree:mainFrameData:]): Deleted.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::updateLayoutViewportParameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::willCommitMainFrameData):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
(WebKit::RemoteLayerTreeDrawingArea::willCommitMainFrameData):
(WebKit::RemoteLayerTreeDrawingArea::willCommitLayerTree): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitMainFrameData):
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree): Deleted.

Canonical link: <a href="https://commits.webkit.org/303348@main">https://commits.webkit.org/303348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2964a02d77a3cbce761fb6352edfa456dbe3f0a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4486 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139508 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83896 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4248 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100898 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3156 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118222 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81689 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ae5f1259-29f7-40cb-8269-324b42bb06e7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3050 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/917 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82726 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111803 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142153 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4156 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36917 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109268 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4237 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3692 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109440 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27746 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3147 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114493 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57377 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4209 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32889 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4041 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67656 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4301 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4169 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1132,26 +1133,28 @@ - (BOOL)_restoreScrollAndZoomStateForTransaction:(const WebKit::RemoteLayerTreeT
-    if (layerTreeTransaction.isMainFrameProcessTransaction()) {
-        [self _trackTransactionCommit:layerTreeTransaction];
+    if (mainFrameData) {
+        const auto& mainFrameCommitData = *mainFrameData;
```
If we have a bundle containing multiple transactions, then this code block is going to run multiple times, right?

I guess that's not new to this PR, but does seem like a problem.

Might be good to have a followup (or a bug filed at least) to split the `didCommitLayerTree` tree of functions into 3 different ones (transaction, page, mainframe), that each get called once per object.

ruthvikkonda replied:
Yeah, unfortunately this block here is one of the only ones that cannot be easily split up. I tried splitting it up to achieve the similar didCommitMainFrameData() pattern that I did with others, but the logic in this function, especially the scrolling logic, is very tightly woven into being per transaction (aka being done inside the transaction loop). Numerous attempts to split it up broke a lot of tests repeatedly. I think this work is out of scope for this radar at least. I feel that someone with a deeper understanding of how the rendering works here should rearchitect it to accommodate the new structure.

I'll file a new bug for it!

ruthvikkonda replied:
https://bugs.webkit.org/show_bug.cgi?id=302892

ruthvikkonda committed:
Organize remaining RemoteLayerTreeTransaction members
https://bugs.webkit.org/show_bug.cgi?id=302267
rdar://164407141

Reviewed by Matt Woodrow.

Right now RemoteLayerTreeDrawingAreaProxy::commitLayerTree currently receives a Vector<std::pair<RemoteLayerTreeTransaction,
RemoteScrollingCoordinatorTransaction>>, one pair per root frame. However, some aspects of the transaction are
scoped to the root frame. Some are scoped to the page. And some are scoped to the main frame. We want to accordingly
introduce some separation/organization here to reduce duplication and remove unnecessary isMainFrameProcessTransaction() checks.

This PR builds on https://github.com/WebKit/WebKit/pull/52671 and https://github.com/WebKit/WebKit/pull/52839. In this PR, we move
the last relevant members from transaction scope to main frame, page, and bundle scope accordingly. WCP propagation
is adjusted to fit this. And where appropriate, processing on the UIProcess side is separated to ensure that mainframe specific
data is only processed once rather than per transaction.

We should only be using bundle.mainFrameData now as the sole source of truth for checking for main frame identity, so isMainFrameProcessTransaction()
is removed. Furthermore, since these members got migrated, some logging was removed from RemoteLayerTreeTransaction. We add it back
in RemoteLayerTreeCommitBundle.mm.

No new functionality is added. This is a simple refactoring. Existing test coverage is enough.
As a result, no new tests are added.

* Source/WebCore/platform/ProcessQualified.h:
(WebCore::ProcessQualified::lessThanSameProcess const):
(WebCore::ProcessQualified::lessThanOrEqualSameProcess const):
(WebCore::ProcessQualified::greaterThanSameProcess const):
(WebCore::ProcessQualified::greaterThanOrEqualSameProcess const):
(WebCore::ProcessQualified::lessThanSameProcess): Deleted.
(WebCore::ProcessQualified::lessThanOrEqualSameProcess): Deleted.
(WebCore::ProcessQualified::greaterThanSameProcess): Deleted.
(WebCore::ProcessQualified::greaterThanOrEqualSameProcess): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.mm: Added.
(WebKit::PageData::description const):
(WebKit::MainFrameData::description const):
(WebKit::RemoteLayerTreeCommitBundle::description const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
(WebKit::RemoteLayerTreeTransaction::RemoteLayerTreeTransaction): Deleted.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTreeDuringAnimatedResize:mainFrameData:]):
(-[WKWebView _trackTransactionCommit:]):
(-[WKWebView _updateScrollViewForTransaction:]):
(-[WKWebView _restoreScrollAndZoomStateForTransaction:]):
(-[WKWebView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(-[WKWebView _didCommitLayerTreeDuringAnimatedResize:]): Deleted.
(-[WKWebView _didCommitLayerTree:mainFrameData:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
(WebKit::PageClientImplCocoa::didCommitMainFrameData):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLayerTree):
(WebKit::PageClientImpl::didCommitMainFrameData):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(-[WKContentView _didCommitLayerTree:mainFrameData:]): Deleted.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::updateLayoutViewportParameters):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::willCommitMainFrameData):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
(WebKit::RemoteLayerTreeDrawingArea::willCommitMainFrameData):
(WebKit::RemoteLayerTreeDrawingArea::willCommitLayerTree): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitMainFrameData):
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree): Deleted.

Canonical link: https://commits.webkit.org/303348@main
